### PR TITLE
Storage Queue

### PR DESCRIPTION
**Pull request by mateuszaaa**
- StorageQueue - double buffered queue persited in runtime storage
- Align malicious collator detection with StorageQueue
- Implement Overhead benchmarks for Ver block execution
- fix frame_executive tests
- use AURA key type for generation & validation of shuffling seed
- take into account bytes consumed by digest log

### CODE REVIEW

File: client/basic-authorship-ver/src/basic_authorship.rs [hunk 1]
```diff
@@ -400,24 +470,48 @@ where
...
-			futures_timer::Delay::new(deadline.saturating_duration_since((self.now)()) / 16).fuse();
 
-		block_builder.apply_previous_block_extrinsics(seed.clone());
+		let mut block_size = 0;
+
+		let get_current_time = &self.now;
+		let is_expired = || get_current_time() > queue_processing_deadline;
+
+		let block_size_limit = block_size_limit.unwrap_or(self.default_block_size_limit);
+		block_builder.apply_previous_block_extrinsics(
```
Comment (mateuszaaa) [inline client/basic-authorship-ver/src/basic_authorship.rs:491]: fetching previous block txs up to half of the slot limit

File: client/block-builder-ver/src/lib.rs [hunk 2]
```diff
@@ -223,17 +221,84 @@ where
...
+
+		apply_transaction_wrapper::<Block, A>(
+			&self.api,
+			&self.block_id,
+			store_txs_inherent.clone(),
+			ExecutionContext::BlockConstruction,
+		)
+		.unwrap()
+		.unwrap()
+		.unwrap();
```
Comment (mateuszaaa) [inline client/block-builder-ver/src/lib.rs:299]: put all txs collected from transaction pool into inherent that stores them in storage queue

File: client/block-builder-ver/src/lib.rs [hunk 3]
```diff
@@ -331,74 +368,68 @@ where
 	/// fetch previous block and apply it
 	///
 	/// consequence of delayed block execution
-	pub fn apply_previous_block_extrinsics(&mut self, seed: ShufflingSeed) {
-		let parent_hash = self.parent_hash;
+	pub fn apply_previous_block_extrinsics<F>(
```
Comment (mateuszaaa) [inline client/block-builder-ver/src/lib.rs:371]: fetch list of all txs from storage queue and then try apply as many as posiible

File: client/block-builder-ver/src/lib.rs [hunk 4]
```diff
@@ -331,74 +368,68 @@ where
...
+				}else{
+					break;
 				}
-			})
+			} else {
+				log::warn!(target: "block_builder", "cannot decode tx");
+			}
 		}
+
+		self.api.pop_txs(&block_id, extrinsics.len() as u64).unwrap();
```
Comment (mateuszaaa) [inline client/block-builder-ver/src/lib.rs:431]: pop executed txs from storage queue (so far we were just executing them)

File: frame/system/src/lib.rs [hunk 5]
```diff
@@ -578,11 +617,51 @@ pub mod pallet {
...
+	/// Its worth to notice that storage queue adds only single storage write, as list of all txs
+	/// is stored as single value (encoded list of txs) maped to single key (block number)
+	///
+	/// # Storage Qeueue interaction
+	/// There are two ways to interact with storage queue:
+	/// - enqueuing new txs using [`Pallet::enqueue_txs`] inherent
+	/// - poping txs from the queue using [`Pallet::pop_txs`] that is exposed throught RuntimeApi
+	///   call
+	#[pallet::storage]
+	pub type StorageQueue<T: Config> = StorageValue<
```
Comment (mateuszaaa) [inline frame/system/src/lib.rs:648]: storage queue entry in runtime storage

File: frame/system/src/lib.rs [hunk 6]
```diff
@@ -387,6 +393,35 @@ pub mod pallet {
...
 		}
 
+		/// Persists list of encoded txs into the storage queue. There is an dedicated
+		/// check in [Executive](https://storage.googleapis.com/mangata-docs-node/frame_executive/struct.Executive.html) that verifies that passed binary data can be
+		/// decoded into extrinsics.
+		#[pallet::weight((
+			0,
+			DispatchClass::Mandatory
+		))]
+		pub fn enqueue_txs(
```
Comment (mateuszaaa) [inline frame/system/src/lib.rs:403]: pushes passed txs into storage queue, also records txs hash in digest log so they can be removed from local copy of transaction pool when block is imported

File: frame/system/src/lib.rs [hunk 7]
```diff
@@ -1299,8 +1378,123 @@ impl<T: Config> Pallet<T> {
...
+				_ => None,
+			})
+			.flatten()
+			.cloned()
+			.collect::<Vec<_>>()
+	}
+
+	/// Dequeue particular number of txs from storage queue.
+	/// It modifies the storage
+	pub fn pop_txs(mut len: usize) -> Vec<EncodedTx> {
```
Comment (mateuszaaa) [inline frame/system/src/lib.rs:1457]: interacts with storage queue, pops specified number of txs

File: frame/system/src/lib.rs [hunk 8]
```diff
@@ -578,11 +617,51 @@ pub mod pallet {
...
+		BoundedVec<
+			(T::BlockNumber, Option<u32>, Vec<(Option<T::AccountId>, EncodedTx)>),
+			StorageQueueLimit,
+		>,
+		ValueQuery,
+	>;
+
+	/// Map of block numbers to block shuffling seeds
+	#[pallet::storage]
+	pub type DidStoreTxs<T: Config> = StorageValue<_, bool, ValueQuery>;
```
Comment (mateuszaaa) [inline frame/system/src/lib.rs:659]: some internal information that eventually is not stored into runtime storage

File: frame/executive/src/lib.rs [hunk 9]
```diff
@@ -495,43 +542,64 @@ where
...
 
-			// verify that all extrinsics can be executed in single block
 			let max = System::BlockWeights::get();
 			let mut all: frame_system::ConsumedWeight = Default::default();
-			for tx in curr_block_txs.clone() {
-				let info = tx.get_dispatch_info();
-				all = frame_system::calculate_consumed_weight::<CallOf<Block::Extrinsic, Context>>(max.clone(), all, &info)
-					.expect("sum of extrinsics should fit into single block");
-			}
+			if let Some((nr, _index, txs)) = frame_system::StorageQueue::<System>::get().last() {
```
Comment (mateuszaaa) [inline frame/executive/src/lib.rs:577]: validate weight of enqueued txs

File: frame/executive/src/lib.rs [hunk 10]
```diff
@@ -541,10 +609,22 @@ where
...
+		sp_runtime::runtime_logger::RuntimeLogger::init();
+		extrinsics.into_iter().for_each(|tx| {
+			let tx_hash = BlakeTwo256::hash(&tx.encode());
+			let is_extrinsic = tx.is_signed().unwrap();
+			if let Err(e) = Self::apply_extrinsic(tx) {
+				log::debug!(target: "runtime::ver", "executing extrinsic :{:?}", tx_hash);
+				// there will be some cases when tx execution may fail (because of delayed execution) so we want to panic only when:
+				// - tx is inherent
+				// - tx is extrinsic and error cause is exhaust resources
+				if !is_extrinsic || matches!(e, TransactionValidityError::Invalid(err) if err.exhausted_resources())
```
Comment (mateuszaaa) [inline frame/executive/src/lib.rs:621]: we allow some txs from queue to fail in some corner cases that we may not be aware

File: client/basic-authorship-ver/src/basic_authorship.rs (hunk 1 above)
Comment (iStrike7) [reply client/basic-authorship-ver/src/basic_authorship.rs:491]: Why is it applying previous block extrinsics, when it should be executing them from queue?

**Review by iStrike7 — APPROVED** [review]

File: client/basic-authorship-ver/src/basic_authorship.rs (hunk 1 above)
Comment (mateuszaaa) [reply client/basic-authorship-ver/src/basic_authorship.rs:491]: yeah, ill improve naming - thats exactly what its doing

**Review by majusko — APPROVED** [review]

**Review by vanderian — APPROVED** [review]

**Pull request merged**